src/Adapter/Cart/Comparator/CartProductsComparator.php

```diff
@@ -0,0 +1,210 @@
+     *
+     * @param CartProductUpdate[] $knownUpdates
+     *
+     * @return CartProductUpdate[]
+     */
+    public function getUpdatedProducts(array $knownUpdates = []): array
+    {
+        $newProducts = $this->cart->getProducts(true);
+        $allUpdateProducts = $this->getAllUpdatedProducts($newProducts);
+
+        return $this->filterKnownUpdates($allUpdateProducts, $knownUpdates);
+    }
+
+    /**
+     * Returns a list of products that were strictly created (not updated) compared to the state of the cart
+     * when this object was created, it removes the already known modified products provided as argument.
+     *
+     * @param CartProductUpdate[] $knownUpdates
+     *
+     * @return CartProductUpdate[]
+     */
+    public function getAdditionalProducts(array $knownUpdates = []): array
+    {
+        $newProducts = $this->cart->getProducts(true);
+        $allAdditionalProducts = $this->getAllAdditionalProducts($newProducts);
+
+        return $this->filterKnownUpdates($allAdditionalProducts, $knownUpdates);
+    }
+
+    /**
+     * Returns a list of products that were modified (created and/or updated) compared to the state of the cart
+     * when this object was created, it removes the already known modified products provided as argument.
+     *
+     * @param CartProductUpdate[] $knownUpdates
+     *
+     * @return CartProductUpdate[]
+     */
+    public function getModifiedProducts(array $knownUpdates = []): array
+    {
+        $newProducts = $this->cart->getProducts(true);
+        $modifiedProducts = array_merge($this->getAllUpdatedProducts($newProducts), $this->getAllAdditionalProducts($newProducts));
+
+        return $this->filterKnownUpdates($modifiedProducts, $knownUpdates);
+    }
+
+    /**
+     * Returns the list of updates for products that were not in the cart previously
+     *
+     * @param array[] $newProducts
+     *
+     * @return array
+     */
+    private function getAllAdditionalProducts(array $newProducts): array
+    {
+        $additionalProducts = [];
+        foreach ($newProducts as $newProduct) {
+            // Then try and find the product in new products
+            $oldProduct = $this->getMatchingProduct($this->savedProducts, $newProduct);
+            if (null === $oldProduct) {
+                $additionalProducts[] = new CartProductUpdate(
+                    (int) $newProduct['id_product'],
+                    (int) $newProduct['id_product_attribute'],
+                    (int) $newProduct['cart_quantity'],
+                    true
+                );
+            }
+        }
+
+        return $additionalProducts;
+    }
+
+    /**
+     * Returns a list of all products that were updated compared to the creation of this object.
+     *
+     * @param array[] $newProducts
+     *
+     * @return CartProductUpdate[]
+     */
+    private function getAllUpdatedProducts(array $newProducts): array
+    {
+        $updatedProducts = [];
```

Comment:
```suggestion
        $updatedProducts = [];
        $deltaQuantity = 0;
```

Reply:
Not useful since `$deltaQuantity` is correctly reset in all loop iterations (if/else) No risk to keep a former value

Reply:
My bad, I read `$deltaQuantity -= (int) $oldProduct['cart_quantity'];` instead of `$deltaQuantity = -(int) $oldProduct['cart_quantity'];` 🤦‍♂️
We're all good then!